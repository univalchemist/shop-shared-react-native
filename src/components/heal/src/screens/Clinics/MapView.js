import React, { useState } from 'react';
import { Box } from '@cxa-rn/components';
import styled from 'styled-components/native';
import { PanelClusteredMapView } from '@heal/src/screens/Clinics/widgets';
import ClinicCard from '@heal/src/screens/Clinics/widgets/clinicCards/ClinicCard';
export const MapViewContainer = styled.View`
  height: 70%;
  width: 100%;
`;

export const MapView = ({ navigation, Index }) => {
  const [mapViewLayout, setMapViewLayout] = useState(undefined);
  return (
    <Box flex={1}>
      {Index === 0 && (
        <MapViewContainer
          onLayout={event => setMapViewLayout(event.nativeEvent.layout)}
        >
          {mapViewLayout && <PanelClusteredMapView dimension={mapViewLayout} />}
        </MapViewContainer>
      )}

      <Box height={'30%'}>
        <ClinicCard navigation={navigation} />
      </Box>
    </Box>
  );
};
