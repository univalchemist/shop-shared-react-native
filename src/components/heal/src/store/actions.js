import * as types from './types';
import moment from 'moment';

export const getDoctors = specialityCode => (dispatch, getState, { api }) => {
  const getPromise = async () => {
    const { clientId } = getState().user;
    const { doctorData } = getState().heal;
    let page = doctorData.page + 1;
    if (doctorData.specialityCode !== specialityCode) {
      page = 1;
    }
    const { data } = await api.getDoctors({ clientId, page, specialityCode });
    return { ...data, specialityCode };
  };
  return dispatch({
    type: types.GET_DOCTORS,
    payload: getPromise(),
  });
};

export const getClinics = ({
  searchTerm,
  searchBy,
  forceReload,
} = {}) => async (dispatch, getState, { api }) => {
  const getPromise = async () => {
    const { clientId } = getState().user;
    const { location, clinicData } = getState().heal;
    let page = clinicData.page + 1;
    if (
      searchTerm !== clinicData.searchTerm ||
      searchBy !== clinicData.searchBy
    )
      page = 1;

    if (forceReload) page = 1;

    const response = await api.getClinicsWithObject({
      clientId,
      lat: location.latitude,
      lng: location.longitude,
      page,
      searchTerm,
      searchBy,
    });
    response.data.items = transformClinicForCluster(response.data.items);
    return { ...response.data, searchTerm, searchBy, forceReload };
  };

  return dispatch({
    type: types.FETCH_CLINICS,
    payload: getPromise(),
  });
};

export const fetchLocation = () => async (dispatch, getState, { api }) => {
  const position = {
    coords: {
      latitude: 0,
      longitude: 0,
    },
  };
  const getPromise = async () => {
    return new Promise(resolve => {
      navigator.geolocation.getCurrentPosition(
        position => resolve(position),
        error => resolve(position),
      );
    });
  };

  return dispatch({
    type: types.FETCH_LOCATION,
    payload: getPromise(),
  });
};

export const getSpecialities = () => (dispatch, getState, { api }) => {
  const getPromise = async () => {
    const { clientId } = getState().user;
    const { data } = await api.getSpecialities(clientId);
    return data;
  };
  return dispatch({
    type: types.GET_SPECIALITIES,
    payload: getPromise(),
  });
};

export const getDoctorsInfo = clinic => (dispatch, getState, { api }) => {
  const getPromise = async () => {
    const { clientId, userId } = getState().user;
    const clinicProviderId = clinic.clinicProviderId;
    const clinicId = clinic.id;
    let promiseArray = [];

    for (let i = 0; i < clinic.doctors.length; i += 1) {
      const doctor = clinic.doctors[i];
      promiseArray.push(
        api.getDoctorInfo({
          clientId,
          userId,
          clinicProviderId,
          clinicId,
          doctorId: doctor.id,
        }),
      );
    }
    const data = await Promise.all(promiseArray);
    return data;
  };
  return dispatch({
    type: types.GET_DOCTORS_INFO,
    payload: getPromise(),
  });
};

export const setDetailsClinic = clinic => (dispatch, getState, { api }) => {
  return dispatch({
    type: types.SET_DETAILS_CLINIC,
    payload: clinic,
  });
};

export const search = ({ searchTerm, latitude, longitude }) => (
  dispatch,
  getState,
  { api },
) => {
  const searchClinics = async ({
    searchBy,
    searchTerm,
    searchType,
    lat = null,
    lng = null,
    nearBy = null,
  }) => {
    const { clientId } = getState().user;
    const { data } = await api.getClinicsWithObject({
      clientId,
      lat,
      lng,
      page: 1,
      itemPerPage: types.GET_ALL,
      nearBy,
      searchBy,
      searchTerm,
    });
    return { ...data, searchType };
  };

  const searchDoctors = async ({
    searchBy,
    searchTerm,
    specialityCode,
    searchType,
  }) => {
    const { clientId } = getState().user;
    const { data } = await api.getDoctors({
      clientId,
      page: 1,
      itemPerPage: types.GET_ALL,
      searchBy,
      searchTerm,
      specialityCode,
    });
    if (specialityCode) return { ...data, specialityCode, searchType };
    else return { ...data, searchType };
  };

  const doSearch = async () => {
    const { specialities, specialitiesByCode } = getState().heal;

    let data = [];
    let raw;
    let specialitySearchResult = [];

    if (searchTerm === types.SearchItemType.NearMe) {
      raw = await searchClinics({
        searchType: types.SearchItemType.NearMe,
        nearBy: types.UNIFY_SEARCH_DISTANCE,
        lat: latitude,
        lng: longitude,
      });
    } else {
      const filteredSpecialities = specialities.filter(s =>
        s.name.toLowerCase().includes(searchTerm.toLowerCase()),
      );

      raw = await Promise.all([
        searchDoctors({
          searchTerm,
          searchBy: types.DoctorSearchType.Name,
          searchType: types.SearchItemType.Doctor,
        }),
        ...filteredSpecialities.map(f =>
          searchDoctors({
            specialityCode: f.code,
            searchType: types.SearchItemType.Speciality,
          }),
        ),
        searchClinics({
          searchBy: types.ClinicSearchType.Name,
          searchTerm,
          searchType: types.SearchItemType.Clinic,
        }),
        searchClinics({
          searchBy: types.ClinicSearchType.Location,
          searchTerm,
          searchType: types.SearchItemType.Location,
        }),
      ]);
    }

    for (let i = 0; i < raw.length; i++) {
      const res = raw[i];
      switch (res.searchType) {
        case types.SearchItemType.Doctor:
          if (res.items?.length > 0)
            data.push({
              title: 'heal.UnifySearch.DoctorName',
              data: res.items.map(d => ({
                title: d.name,
                subtitle: null,
                data: { ...d, type: types.SearchItemType.Doctor },
              })),
            });
          break;
        case types.SearchItemType.Speciality:
          if (res.totalCount > 0) {
            specialitySearchResult.push({
              title: specialitiesByCode[res.specialityCode].name,
              subtitle: res.totalCount,
              data: {
                specialityCode: res.specialityCode,
                type: types.SearchItemType.Speciality,
              },
            });
          }
          const s = data.find(d => d.title === 'heal.UnifySearch.Speciality');
          if (!s && specialitySearchResult.length > 0)
            data.push({
              title: 'heal.UnifySearch.Speciality',
              data: specialitySearchResult,
            });
          break;
        case types.SearchItemType.Location:
          if (res.totalCount > 0) {
            let location = {};
            const items = res.items;
            for (let i = 0; i < items.length; i++) {
              if (!location[items[i].district]) location[items[i].district] = 0;
              location[items[i].district] += 1;
            }
            data.push({
              title: 'heal.UnifySearch.Location',
              data: Object.keys(location).map(k => ({
                title: k,
                subtitle: location[k],
                data: {
                  districtCode: k,
                  type: types.SearchItemType.Location,
                },
              })),
            });
          }
          break;
        case types.SearchItemType.Clinic:
          if (res.items?.length > 0)
            data.push({
              title: 'heal.UnifySearch.ClinicName',
              data: res.items.map(c => ({
                title: c.name,
                subtitle: null,
                data: { ...c, type: types.SearchItemType.Clinic },
              })),
            });
          break;
      }
    }

    console.log(data);
    return data;
  };

  return dispatch({
    type: types.SEARCH,
    payload: doSearch(),
  });
};

export const clearSearch = () => (dispatch, getState, { api }) => {
  return dispatch({ type: types.CLEAR_SEARCH });
};

const transformClinicForCluster = clinics => {
  const data = clinics.map(clinic => ({
    ...clinic,
    location: { longitude: clinic.longitude, latitude: clinic.latitude },
  }));
  return data;
};

export const updateSelectedClinic = clinic => dispatch => {
  return dispatch({
    type: types.UPDATE_SELECTED_CLINIC,
    payload: clinic,
  });
};

export const updateSelectedClinics = clinics => dispatch => {
  return dispatch({
    type: types.UPDATE_SELECTED_CLINICS,
    payload: clinics,
  });
};

export const updateClinicPage = () => dispatch => {
  return dispatch({
    type: types.UPDATE_CLINIC_PAGE,
  });
};

export const scanQRCode = clinicQrCode => (dispatch, getState, { api }) => {
  const scan = async () => {
    const { clientId, userId } = getState().user;
    const { data } = await api.scanQRCode({ clientId, userId, clinicQrCode });
    return { ...data, clinicQrCode };
  };
  return dispatch({
    type: types.SCAN_QR_CODE,
    payload: scan(),
  });
};

export const checkInWalkIn = ({ clinicProviderId, clinicId, doctorId }) => (
  dispatch,
  getState,
  { api },
) => {
  const getPromise = async () => {
    const { clientId, userId } = getState().user;
    const { data } = api.checkInWalkIn({
      clientId,
      userId,
      clinicProviderId,
      clinicId,
      doctorId,
    });
    return data;
  };

  return dispatch({
    type: types.CHECK_IN_WALK_IN,
    payload: getPromise(),
  });
};

export const getWalkInSelectDoctors = clinicQrCode => (
  dispatch,
  getState,
  { api },
) => {
  const getPromise = async () => {
    const { clientId, userId } = getState().user;
    const { data } = await api.getWalkInSelectDoctors({
      clientId,
      userId,
      clinicQrCode,
    });
    return data;
    //     return {
    //       clinic: {
    //         id: 373,
    //         name: 'CXA (SG) Testing Clinic 1',
    //         latitude: 1.303007,
    //         longitude: 103.796334,
    //         qrCode: '304',
    //         district: 'CAUSEWAY BAY',
    //         area: null,
    //       },
    //       doctors: [
    //         {
    //           isOpen: false,
    //           isRemoteFull: false,
    //           isRemoteTicketEnabled: true,
    //           isAppointmentEnabled: true,
    //           position: 0,
    //           estimatedConsultationTime: '1:48 pm',
    //           ticketDepositAmount: 0.0,
    //           canGetRefundOnCancellation: false,
    //           cancellationGraceMinutes: 0,
    //           id: 355,
    //           name: 'Chan Tai Man',
    //           specialityCode: 'generalpractitio',
    //           gender: 'Male',
    //           email: null,
    //           imageUrl: null,
    //           introduction: null,
    //         },
    //         {
    //           isOpen: false,
    //           isRemoteFull: false,
    //           isRemoteTicketEnabled: true,
    //           isAppointmentEnabled: true,
    //           position: 0,
    //           estimatedConsultationTime: '1:48 pm',
    //           ticketDepositAmount: 0.0,
    //           canGetRefundOnCancellation: false,
    //           cancellationGraceMinutes: 0,
    //           id: 354,
    //           name: 'CXA2',
    //           specialityCode: 'generalpractitio',
    //           gender: 'Male',
    //           email: null,
    //           imageUrl: null,
    //           introduction: null,
    //         },
    //       ],
    //     };
  };

  return dispatch({
    type: types.GET_WALKIN_SELECTS,
    payload: getPromise(),
  });
};

export const getMedicalProfile = () => (dispatch, getState, { api }) => {
  const fetchProfile = async () => {
    const { clientId } = getState().user;
    const {
      values: { memberId },
    } = getState().form.checkinForm;
    const { data } = await api.getMedicalProfile(clientId, memberId);
    return data;
  };
  return dispatch({
    type: types.GET_MEDICAL_PROFILE,
    payload: fetchProfile(),
  });
};

export const updateMedicalProfile = () => (dispatch, getState, { api }) => {
  const updateProfile = async () => {
    const { clientId } = getState().user;
    const { values } = getState().form.checkinForm;
    const { data } = await api.updateMedicalProfile({
      clientId,
      userId: values.memberId,
      ...values,
    });
    return data;
  };
  return dispatch({
    type: types.UPDATE_MEDICAL_PROFILE,
    payload: updateProfile(),
  });
};

export const getAppointmentList = () => (dispatch, getState, { api }) => {
  const fetchAppointment = async () => {
    const { clientId, userId } = getState().user;
    const { data } = await api.getAppointmentList(clientId, userId);
    dispatch({
      type: types.GET_APPOINTMENT_LIST,
      payload: data,
    });
  };
  fetchAppointment();
};

export const acceptAppointment = (confirm, appointmentId) => (
  dispatch,
  getState,
  { api },
) => {
  const updateAppointment = async () => {
    const { clientId, userId } = getState().user;
    const { data } = await api.acceptAppointment({
      clientId,
      userId: userId,
      appointmentId,
      isConfirmed: confirm,
    });
    return data;
  };
  return dispatch({
    type: types.ACCEPT_APPOINTMENT,
    payload: updateAppointment(),
  });
};

export const getRemoteTickets = () => (dispatch, getState, { api }) => {
  const getTickets = async () => {
    const { clientId, userId } = getState().user;
    const { data } = await api.getRemoteTickets(clientId, userId);
    return data;
  };
  return dispatch({
    type: types.GET_REMOTE_TICKETS,
    payload: getTickets(),
  });
};

export const getMembers = () => (dispatch, getState, { api }) => {
  const getPromise = async () => {
    const { clientId, userId } = getState().user;
    const response = await api.fetchMemberProfile(clientId, userId);
    return response.data;
  };
  return dispatch({
    type: types.GET_MEMBER,
    payload: getPromise(),
  });
};

export const requestAppointment = (
  clinicProviderId,
  doctorId,
  clinicId,
  dates,
) => (dispatch, getState, { api }) => {
  const getPromise = async () => {
    const { clientId, userId } = getState().user;
    let datesObj = {};
    dates.map((item, index) => {
      if (index === 0) {
        datesObj.firstPreferredTime = moment(item.date).format();
        datesObj.firstSession = item.time;
      }
      if (index === 1) {
        datesObj.secondPreferredTime = moment(item.date).format();
        datesObj.secondSession = item.time;
      }
      if (index === 2) {
        datesObj.thirdPreferredTime = moment(item.date).format();
        datesObj.thirdSession = item.time;
      }
    });
    const appointmentData = {
      clinicProviderId: clinicProviderId,
      doctorId: doctorId,
      clinicId: clinicId,
      ...datesObj,
    };
    const { data } = await api.requestAppointment({
      clientId,
      userId,
      appointmentData,
    });
    return data;
  };
  return dispatch({
    type: types.REQUEST_APPOINTMENT,
    payload: getPromise(),
  });
};
