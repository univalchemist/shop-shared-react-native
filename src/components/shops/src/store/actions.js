/* istanbul ignore file */

export * from './config/actions';
export * from './filters/actions';
export * from './home/actions';
export * from './review/actions';
export * from './cart/actions';
export * from './orderhistory/actions';
export * from './deliveryAddress/actions';
