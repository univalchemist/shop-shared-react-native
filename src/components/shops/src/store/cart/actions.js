import {
  ADD_TO_CART,
  GET_CART,
  GET_CART_BILLING,
  GET_CART_SHIPPING,
  GET_CART_TOTALS,
  REMOVE_ITEM_FROM_CART,
} from './types';

export const addToCart = ({ sku, quantity }) => (
  dispatch,
  getState,
  { api },
) => {
  const getPromise = async () => {
    const { userId, clientId } = getState().user;
    const { data } = await api.addToCart({
      clientId,
      userId,
      sku,
      quantity,
    });
    return data;
  };

  return dispatch({
    type: ADD_TO_CART,
    payload: getPromise(),
  });
};

export const getCart = () => (dispatch, getState, { api }) => {
  const getPromise = async () => {
    const { userId, clientId } = getState().user;
    const { data } = await api.getCart(clientId, userId);
    return data;
  };

  return dispatch({
    type: GET_CART,
    payload: getPromise(),
  });
};

export const getCartTotals = () => (dispatch, getState, { api }) => {
  const getPromise = async () => {
    const { userId, clientId } = getState().user;
    const { data } = await api.getCartTotals(clientId, userId);
    return data;
  };

  return dispatch({
    type: GET_CART_TOTALS,
    payload: getPromise(),
  });
};

export const getCartBilling = () => (dispatch, getState, { api }) => {
  const getPromise = async () => {
    const { userId, clientId } = getState().user;
    const { data } = await api.getCartBilling(clientId, userId);
    return data;
  };

  return dispatch({
    type: GET_CART_BILLING,
    payload: getPromise(),
  });
};

export const getCartShipping = () => (dispatch, getState, { api }) => {
  const getPromise = async () => {
    const { userId, clientId } = getState().user;
    const { data } = await api.getCartShipping(clientId, userId);
    return data;
  };

  return dispatch({
    type: GET_CART_SHIPPING,
    payload: getPromise(),
  });
};

export const removeItemFromCart = sku => (dispatch, getState, { api }) => {
  const getPromise = async () => {
    const { userId, clientId } = getState().user;
    const { data } = await api.removeItemFromCart(clientId, userId, sku);
    return data;
  };

  return dispatch({
    type: REMOVE_ITEM_FROM_CART,
    payload: getPromise(),
  });
};
