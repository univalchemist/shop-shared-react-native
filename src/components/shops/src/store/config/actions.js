import {
  GET_STORE_CONFIGS,
  GET_CATEGORIES,
  GET_SORTINGS,
  GET_COUNTRIES,
} from './types';

export const getStoreConfigs = () => async (dispatch, getState, { api }) => {
  const getPromise = async () => {
    const { clientId } = getState().user;

    const { data } = await api.getStoreConfigs(clientId);
    return data && data[0];
  };

  return dispatch({
    type: GET_STORE_CONFIGS,
    payload: getPromise(),
  });
};

export const getCategories = () => async (dispatch, getState, { api }) => {
  const getPromise = async () => {
    const { clientId } = getState().user;

    const { data } = await api.getCategories(clientId);
    return data && data[0];
  };

  return dispatch({
    type: GET_CATEGORIES,
    payload: getPromise(),
  });
};

export const getSortings = () => async (dispatch, getState, { api }) => {
  const getPromise = async () => {
    const { clientId } = getState().user;

    const { data } = await api.getSortings(clientId);
    return data;
  };

  return dispatch({
    type: GET_SORTINGS,
    payload: getPromise(),
  });
};

export const getCountries = () => async (dispatch, getState, { api }) => {
  const getPromise = async () => {
    const { clientId } = getState().user;

    const { data } = await api.getCountries(clientId);
    return data;
  };

  return dispatch({
    type: GET_COUNTRIES,
    payload: getPromise(),
  });
};
