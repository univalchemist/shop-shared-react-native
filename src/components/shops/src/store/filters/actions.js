import {
  UPDATE_SELECTED_CATEGORIES,
  UPDATE_SORT_TYPE,
  UPDATE_FILTER_TYPES,
  UPDATE_SEARCH_STRING,
} from './types';

export const updateSelectedCategories = categories => ({
  type: UPDATE_SELECTED_CATEGORIES,
  payload: categories,
});

export const updateSortType = sortType => ({
  type: UPDATE_SORT_TYPE,
  payload: sortType,
});

export const updateFilterTypes = filterTypes => ({
  type: UPDATE_FILTER_TYPES,
  payload: filterTypes,
});

export const updateSearchString = searchString => ({
  type: UPDATE_SEARCH_STRING,
  payload: searchString,
});
