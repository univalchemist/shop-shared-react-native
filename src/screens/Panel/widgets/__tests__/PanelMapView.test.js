import React from 'react';
import { Platform, PermissionsAndroid } from 'react-native';
import { renderForTest } from '@testUtils';
import MapView, { Marker } from 'react-native-maps';
import { fireEvent, flushMicrotasksQueue } from 'react-native-testing-library';
import ConnectedPanelMapView, {
  SelectedMapMarker,
  PanelMapView,
} from '../PanelMapView';
import * as actions from '@store/panel/actions';

jest.useFakeTimers();

describe('Panel Map View', () => {
  const defaultLocation = {
    latitude: 22.266021,
    longitude: 114.186422,
    latitudeDelta: 0.2,
    longitudeDelta: 0.2,
  };

  const clinics = [
    {
      id: 1,
      latitude: 4.335,
      longitude: 2.441,
      displayLatitude: 4.3355,
      displayLongitude: 2.4411,
      name: 'Dr John Tan',
    },
    {
      id: 2,
      latitude: 1.555,
      longitude: 0.321,
      displayLatitude: 1.5555,
      displayLongitude: 0.3211,
      name: 'Surgeon Joseline',
    },
    {
      id: 3,
      latitude: 6,
      longitude: 7,
      displayLatitude: 6.6,
      displayLongitude: 7.7,
      name: 'Dr Ian Lin',
    },
  ];

  let originalPlatformOS;
  let originalNavigator;
  let newLocation;

  beforeAll(() => {
    originalPlatformOS = Platform.OS;
    originalNavigator = global.navigator;
    newLocation = {
      coords: {
        longitude: 1.003,
        latitude: 1.4343,
      },
    };
    global.navigator = {
      geolocation: {
        getCurrentPosition: jest.fn(cb => cb(newLocation)),
      },
    };
  });

  afterAll(() => {
    Platform.OS = originalPlatformOS;
    global.navigator = originalNavigator;
    jest.clearAllMocks();
  });

  it('match snapshot', () => {
    const panelMapView = renderForTest(<ConnectedPanelMapView />, {
      initialState: {
        panel: {
          filteredClinics: [clinics[0]],
          defaultLocation,
          selectedClinic: clinics[0],
        },
      },
    });

    const mapView = panelMapView.queryAllByType(MapView)[0];
    mapView.props.onMapReady();

    expect(panelMapView.toJSON()).toMatchSnapshot();
  });

  it('renders the MapView with correct props value', () => {
    const panelMapView = renderForTest(<ConnectedPanelMapView />, {
      initialState: {
        panel: {
          filteredClinics: [clinics[0]],
          defaultLocation,
          selectedClinic: clinics[0],
        },
      },
    });

    expect(panelMapView.queryAllByType(MapView).length).toBe(1);
    const mapView = panelMapView.queryAllByType(MapView)[0];
    mapView.props.onMapReady();
    expect(mapView.props.initialRegion).toBe(defaultLocation);
  });

  it('renders the map markers for all clinics using their display coordinates', () => {
    const panelMapView = renderForTest(<ConnectedPanelMapView />, {
      initialState: {
        panel: {
          filteredClinics: clinics,
          defaultLocation,
          selectedClinic: clinics[0],
        },
      },
    });

    const mapView = panelMapView.queryAllByType(MapView)[0];
    mapView.props.onMapReady();

    const markers = panelMapView.queryAllByType(Marker);
    expect(markers.length).toBe(clinics.length);
    expect(markers[0].props.coordinate).toEqual({
      latitude: clinics[0].displayLatitude,
      longitude: clinics[0].displayLongitude,
    });
    expect(markers[1].props.coordinate).toEqual({
      latitude: clinics[1].displayLatitude,
      longitude: clinics[1].displayLongitude,
    });
    expect(markers[2].props.coordinate).toEqual({
      latitude: clinics[2].displayLatitude,
      longitude: clinics[2].displayLongitude,
    });
  });

  it('should show red marker for selected clinic', async () => {
    const initialState = {
      panel: {
        filteredClinics: clinics,
        defaultLocation,
        selectedClinic: clinics[1],
      },
    };
    const panelMapView = renderForTest(<ConnectedPanelMapView />, {
      initialState,
    });

    const mapView = panelMapView.queryAllByType(MapView)[0];
    mapView.props.onMapReady();

    expect(panelMapView.queryAllByType(SelectedMapMarker).length).toBe(1);
    expect(panelMapView.queryAllByType(SelectedMapMarker)[0].props.item).toBe(
      clinics[1],
    );
  });

  it('should update selectedClinic when user select a pin', async () => {
    jest.spyOn(actions, 'updateSelectedClinic');
    const updateSelectedClinic = jest.fn();
    const initialState = {
      panel: {
        defaultLocation,
        selectedClinic: clinics[0],
      },
    };
    const panelMapView = renderForTest(
      <PanelMapView
        clinics={clinics}
        defaultLocation={defaultLocation}
        selectedClinic={clinics[0]}
        updateSelectedClinic={updateSelectedClinic}
      />,
      {
        initialState,
      },
    );

    const mapView = panelMapView.queryAllByType(MapView)[0];
    mapView.props.onMapReady();

    const markers = panelMapView.queryAllByType(Marker);
    expect(markers.length).toBe(clinics.length);
    const selectedIndex = 2;
    fireEvent.press(markers[selectedIndex]);

    await flushMicrotasksQueue();
    expect(updateSelectedClinic).toHaveBeenCalledWith(clinics[selectedIndex]);
  });

  it('should get current user location when location permission is granted', () => {
    Platform.OS = 'android';

    jest
      .spyOn(PermissionsAndroid, 'request')
      .mockResolvedValue(PermissionsAndroid.RESULTS.GRANTED);

    const initialState = {
      panel: { clinics, defaultLocation, selectedClinic: clinics[0] },
    };

    const updateSelectedClinic = jest.fn();
    renderForTest(
      <PanelMapView
        clinics={clinics}
        defaultLocation={defaultLocation}
        selectedClinic={clinics[0]}
        updateSelectedClinic={updateSelectedClinic}
      />,
      {
        initialState,
      },
    );

    expect(navigator.geolocation.getCurrentPosition).toHaveBeenCalled();
  });

  it('should animate to user location when permission is granted and no clinics available ', async () => {
    Platform.OS = 'android';

    jest
      .spyOn(PermissionsAndroid, 'request')
      .mockResolvedValue(PermissionsAndroid.RESULTS.GRANTED);

    const initialState = {
      panel: {
        filteredClinics: [],
        defaultLocation,
        selectedClinic: clinics[0],
      },
    };

    const updateSelectedClinic = jest.fn();
    const panelMapView = renderForTest(
      <PanelMapView
        clinics={[]}
        defaultLocation={defaultLocation}
        selectedClinic={clinics[0]}
        updateSelectedClinic={updateSelectedClinic}
      />,
      {
        initialState,
      },
    );

    const mapView = panelMapView.queryAllByType(MapView)[0];
    mapView.props.onMapReady();

    await flushMicrotasksQueue();

    jest.runOnlyPendingTimers();
    expect(setTimeout).toHaveBeenCalledWith(expect.any(Function), 1000);
    expect(MapView.prototype.animateToRegion).toHaveBeenCalledWith(
      expect.objectContaining(newLocation.coords),
    );
  });

  it('should animate to show all markers when permission is granted and clinics are available ', async () => {
    Platform.OS = 'android';

    jest
      .spyOn(PermissionsAndroid, 'request')
      .mockResolvedValue(PermissionsAndroid.RESULTS.GRANTED);

    const initialState = {
      panel: {
        filteredClinics: clinics,
        defaultLocation,
        selectedClinic: clinics[0],
      },
    };

    const updateSelectedClinic = jest.fn();
    const panelMapView = renderForTest(
      <PanelMapView
        clinics={clinics}
        defaultLocation={defaultLocation}
        selectedClinic={clinics[0]}
        updateSelectedClinic={updateSelectedClinic}
      />,
      {
        initialState,
      },
    );

    const mapView = panelMapView.queryAllByType(MapView)[0];
    mapView.props.onMapReady();

    await flushMicrotasksQueue();
    expect(MapView.prototype.fitToCoordinates).toHaveBeenCalledWith(
      [
        { latitude: 4.335, longitude: 2.441 },
        { latitude: 1.555, longitude: 0.321 },
        { latitude: 6, longitude: 7 },
      ],
      {
        animated: true,
        edgePadding: { bottom: 20, left: 20, right: 20, top: 20 },
      },
    );
  });

  it('should animate to show one marker using animateToRegion when permission is granted and only one clinic is available ', async () => {
    Platform.OS = 'android';

    jest
      .spyOn(PermissionsAndroid, 'request')
      .mockResolvedValue(PermissionsAndroid.RESULTS.GRANTED);

    const initialState = {
      panel: {
        filteredClinics: clinics,
        defaultLocation,
        selectedClinic: clinics[0],
      },
    };

    const updateSelectedClinic = jest.fn();
    const panelMapView = renderForTest(
      <PanelMapView
        clinics={[clinics[0]]}
        defaultLocation={defaultLocation}
        selectedClinic={clinics[0]}
        updateSelectedClinic={updateSelectedClinic}
      />,
      {
        initialState,
      },
    );

    const mapView = panelMapView.queryAllByType(MapView)[0];
    mapView.props.onMapReady();

    await flushMicrotasksQueue();

    jest.runOnlyPendingTimers();
    expect(setTimeout).toHaveBeenCalledWith(expect.any(Function), 1000);
    expect(MapView.prototype.animateToRegion).toHaveBeenCalledWith({
      ...clinics[0],
      latitudeDelta: 0.02,
      longitudeDelta: 0.02,
    });
  });
});
