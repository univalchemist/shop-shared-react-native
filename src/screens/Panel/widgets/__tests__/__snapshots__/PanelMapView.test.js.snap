// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel Map View match snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaView
    onInsetsChange={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "flex": 1,
          "overflow": "hidden",
        }
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          closing={false}
          gestureVelocityImpact={0.3}
          importantForAccessibility="auto"
          onClose={[Function]}
          onGestureBegin={[Function]}
          onGestureCanceled={[Function]}
          onOpen={[Function]}
          onTransitionStart={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          transitionSpec={
            Object {
              "close": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 0.01,
                  "restSpeedThreshold": 0.01,
                  "stiffness": 1000,
                },
              },
              "open": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 0.01,
                  "restSpeedThreshold": 0.01,
                  "stiffness": 1000,
                },
              },
            }
          }
        >
          <View
            pointerEvents="box-none"
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              enabled={false}
              hitSlop={
                Object {
                  "right": -700,
                }
              }
              maxDeltaY={20}
              minOffsetX={5}
              onHandlerStateChange={[Function]}
            >
              <View
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  pointerEvents="none"
                  style={
                    Object {
                      "backgroundColor": "#fff",
                      "bottom": 0,
                      "left": 0,
                      "position": "absolute",
                      "shadowColor": "#000",
                      "shadowOffset": Object {
                        "height": 1,
                        "width": -1,
                      },
                      "shadowOpacity": 0.3,
                      "shadowRadius": 5,
                      "top": 0,
                      "width": 3,
                    }
                  }
                />
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "height": "100%",
                            "position": "relative",
                          }
                        }
                      >
                        <MapView
                          initialRegion={
                            Object {
                              "latitude": 22.266021,
                              "latitudeDelta": 0.2,
                              "longitude": 114.186422,
                              "longitudeDelta": 0.2,
                            }
                          }
                          loadingEnabled={true}
                          onMapReady={[Function]}
                          showsMyLocationButton={true}
                          showsUserLocation={true}
                          style={
                            Object {
                              "bottom": 0,
                              "left": 0,
                              "margin": 0,
                              "position": "absolute",
                              "right": 0,
                              "top": 0,
                            }
                          }
                        >
                          <Marker
                            coordinate={
                              Object {
                                "latitude": 4.3355,
                                "longitude": 2.4411,
                              }
                            }
                            onPress={[Function]}
                          >
                            <Image
                              source={
                                Object {
                                  "testUri": "../../../src/images/panelClickedMapMarker.png",
                                }
                              }
                              style={
                                Array [
                                  Object {},
                                  Object {
                                    "height": 24,
                                    "width": 24,
                                  },
                                ]
                              }
                            />
                          </Marker>
                        </MapView>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaView>
</View>
`;
